src/Snippet/HtmlInjector.php

```diff
@@ -0,0 +1,294 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Bolt\Snippet;
+
+use Bolt\Common\Str;
+use Bolt\Widget\WidgetInterface;
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * Class for matching HTML elements and injecting text.
+ */
+class HtmlInjector
+{
+    /**
+     * Get a map of function names to locations. We don't have unique callbacks
+     * for all targets, because in practice they end up in about the same
+     * location
+     */
+    public function getMap(): array
+    {
+        return [
+            Target::END_OF_HEAD => 'headTagEnd',
+            Target::AFTER_HEAD_JS => 'headTagEnd', // same as end of head
+            Target::AFTER_HEAD_CSS => 'headTagEnd', // same as end of head
+            Target::AFTER_HEAD_META => 'headTagEnd', // same as end of head because meta tags are unordered
+
+            Target::BEFORE_CSS => 'cssTagsBefore',
+            Target::BEFORE_JS => 'jsTagsBefore',
+            Target::AFTER_META => 'metaTagsAfter',
+            Target::AFTER_CSS => 'cssTagsAfter',
+            Target::AFTER_JS => 'jsTagsAfter',
+
+            Target::START_OF_HEAD => 'headTagStart',
+            Target::BEFORE_HEAD_JS => 'headTagStart', // same as start of head
+            Target::BEFORE_HEAD_CSS => 'headTagStart', // same as start of head
+            Target::BEFORE_HEAD_META => 'headTagStart', // same as start of head because meta tags are unordered
+
+            Target::START_OF_BODY => 'bodyTagStart',
+            Target::BEFORE_BODY_JS => 'bodyTagStart', // same as start of body
+            Target::BEFORE_BODY_CSS => 'bodyTagStart', // same as start of body
+
+            Target::END_OF_BODY => 'bodyTagEnd',
+            Target::AFTER_BODY_JS => 'bodyTagEnd',   // same as end of body
+            Target::AFTER_BODY_CSS => 'bodyTagEnd',   // same as end of body
+
+            Target::END_OF_HTML => 'htmlTagEnd',
+            Target::AFTER_HTML => 'htmlTagEnd',
+
+            Target::NOWHERE => 'nowhere',
+        ];
+    }
+
+    public function inject(WidgetInterface $widget, Response $response): void
+    {
+        $html = $response->getContent();
+        $functionMap = $this->getMap();
+        $target = $widget->getTarget();
+
+        $widget->setResponse($response);
```

Comment:
Only if widget implements ResponseAware (I don't see any Widget that uses response though)

Reply:
Done!

Reply:
(there's one now: The one that insers the `x-powered-by`-header :-)